// e621/build/src/modules/UserFeedback.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const FormHelper_1 = __importDefault(require("../util/FormHelper"));
const RequestHandler_1 = require("../util/RequestHandler");
class UserFeedback {
    auth;
    main;
    constructor(main) {
        Object.defineProperties(this, {
            auth: {
                get() {
                    return !this.main.options.authUser || !this.main.options.authKey ? null : `Basic ${Buffer.from(`${this.main.options.authUser}:${this.main.options.authKey}`).toString("base64")}`;
                },
                configurable: false,
                enumerable: false
            },
            main: {
                value: main,
                configurable: false,
                enumerable: false,
                writable: false
            }
        });
    }
    /**
     * Get a specific user feedback
     *
     * @param {number} id - The id of the feedback to fetch
     * @returns {Promise<(UserFeedbackProperties | null)>}
     */
    async get(id) {
        const res = await this.main.request.get(`/user_feedbacks/${id}.json`).catch(err => {
            if (err instanceof RequestHandler_1.APIError && err.statusCode === 404)
                return null;
            throw err;
        });
        return res;
    }
    /**
     * Search for user feedback
     *
     * @param {object} [options]
     * @param {string} [options.username] - narrow the search by a specific target
     * @param {string} [options.creator] - narrow the search by a specific creator
     * @param {string} [options.body] - narrow the search by the content of the feedback
     * @param {FeedbackCategories} [options.category] - narrow the search by the type
     * @param {(number |`${"" | "a" | "b"}${number}`)} [page] - page of results to get
     * @param {number} [options.limit] - limit the maximum returned results
     * @returns {Promise<Array<UserFeedbackProperties>>}
     */
    async search(options) {
        options = options ?? {};
        const qs = new FormHelper_1.default();
        if (typeof options.username === "string")
            qs.add("search[user_name]", options.username);
        if (typeof options.creator === "string")
            qs.add("search[creator_name]", options.creator);
        if (typeof options.body === "string")
            qs.add("search[body_matches]", options.body);
        if (typeof options.category === "string")
            qs.add("search[category]", options.category);
        if (typeof options.page !== "undefined")
            qs.add("page", options.page);
        if (typeof options.limit === "number")
            qs.add("limit", options.limit);
        const res = await this.main.request.get(`/user_feedbacks.json?${qs.build()}`);
        if (res && !Array.isArray(res) && "user_feedbacks" in res)
            return [];
        return res;
    }
    /**
     * Add feedback to a user
     *
     * * Requires Authentication
     *
     * * Requires Moderator
     *
     * @param {object} options
     * @param {string} options.username - the user the feedback is for
     * @param {FeedbackCategories} options.category - the type of feedback
     * @param {string} options.body - the content of the feedback
     * @returns {Promise<UserFeedbackProperties>}
     */
    async create(options) {
        this.main.request.authCheck.call(this, "UserFeedback#create");
        const qs = new FormHelper_1.default()
            .add("user_feedback[user_name]", options.username)
            .add("user_feedback[category]", options.category)
            .add("user_feedback[body]", options.body);
        const res = await this.main.request.post("/user_feedbacks.json", qs.build());
        return res;
    }
    /**
     * Modify an existing user feedback
     *
     * * Requires Authentication
     *
     * * Requires Moderator
     *
     * @param {number} id - the id of the feedback to edit
     * @param {object} options
     * @param {string} [options.body] - the content of the feedback
     * @param {FeedbackCategories} [options.category] - the type of feedback
     * @returns {Promise<UserFeedbackProperties>}
     */
    async modify(id, options) {
        this.main.request.authCheck.call(this, "UserFeedback#modify");
        const qs = new FormHelper_1.default();
        if (options.body)
            qs.add("user_feedback[body]", options.body);
        if (options.category)
            qs.add("user_feedback[category]", options.category);
        const res = await this.main.request.patch(`/user_feedbacks/${id}.json`, qs.build());
        return res;
    }
    /**
     * Delete an existing user feedback
     *
     * * Requires Authentication
     *
     * * Requires Moderator
     *
     * @param {number} id - the id of the feedback to delete
     * @returns {Promise<null>}
     */
    async delete(id) {
        this.main.request.authCheck.call(this, "UserFeedback#delete");
        const res = await this.main.request.delete(`/user_feedbacks/${id}.json`);
        return res;
    }
}
exports.default = UserFeedback;
