// yiff/build/src/request.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = __importDefault(require("axios"));
const consts_1 = __importDefault(require("./consts"));
async function request(options) {
    switch (options.site) {
        case 'e621':
            if (!options.tags)
                throw Error("No Tags provided");
            let e6request = await axios_1.default({
                method: 'get',
                url: options.killswitch?.enabled ? `${options.killswitch.instance}${consts_1.default.killswitch.e621}?limit=${options.limit || 1}&tags=${options.tags}&useragent=${options.useragent}` :
                    `${consts_1.default.direct.e621}?tags=limit:${options.limit || 1} order:random -young ${options.tags.toString()}`,
                headers: {
                    "User-Agent": options.useragent,
                    // @ts-ignore
                    ...(options?.apikey?.e621 ? {
                        // @ts-ignore
                        "Authorization": options?.apikey?.e621
                    } : {})
                }
            });
            return e6request.data;
        case 'e926':
            if (!options.tags)
                throw Error("No Tags provided");
            let e9request = await axios_1.default({
                method: 'get',
                url: options.killswitch?.enabled ? `${options.killswitch.instance}${consts_1.default.killswitch.e926}?limit=${options.limit || 1}&tags=${options.tags}&useragent=${options.useragent}` :
                    `${consts_1.default.direct.e926}?tags=limit:${options.limit || 1} order:random -young ${options.tags.toString()}`,
                headers: {
                    "User-Agent": options.useragent,
                    // @ts-ignore
                    ...(options?.apikey?.e926 ? {
                        // @ts-ignore
                        "Authorization": options?.apikey?.e926
                    } : {})
                }
            });
            return e9request.data;
        case 'gelbooru':
            if (!options.tags)
                throw Error("No Tags provided");
            let gelboorureq = await axios_1.default({
                method: 'get',
                url: options.killswitch?.enabled ? `${options.killswitch.instance}${consts_1.default.killswitch.gelbooru}?limit=${options.limit || 1}&tags=${options.tags}&useragent=${options.useragent}` :
                    `${consts_1.default.direct.gelbooru}/index.php?page=dapi&s=post&json=1&q=index&limit=${Number(options.limit) || 1}&tags=sort:random+${options.tags.toString().split(' ').join("+")}`,
                headers: {
                    "User-Agent": options.useragent,
                    // @ts-ignore
                    // ...(options?.apikey?.e926 ? {
                    //     // @ts-ignore
                    //     "Authorization": options?.apikey?.e926
                    // } : {})
                }
            });
            return gelboorureq.data;
        case 'furrybot':
        case 'yiffrest':
            let customMethod;
            if (options.killswitch?.enabled) {
                customMethod = "POST";
            }
            else {
                customMethod = "GET";
            }
            let yiffreq = await axios_1.default({
                method: customMethod,
                url: options.killswitch?.enabled ? `${options.killswitch.instance}${consts_1.default.killswitch.yiffrest}` : `${consts_1.default.direct.yiffrest}/${options.category}/${options.endpoint}`,
                data: {
                    category: options.category,
                    endpoint: options.endpoint,
                    apikey: options.apikey,
                    useragent: options.useragent
                },
                headers: {
                    "User-Agent": options.useragent,
                    // @ts-ignore
                    ...(options.apikey?.yiffrest ? {
                        // @ts-ignore
                        "Authorization": options?.apikey?.yiffrest
                    } : {})
                }
            });
            return yiffreq.data;
        case 'sheri':
            if (options.apikey) {
                // @ts-ignore
                if (!options.apikey.startsWith('Token')) {
                    // @ts-ignore
                    options.apikey = `Token ${options.apikey}`;
                }
            }
            let sherireq = await axios_1.default({
                method: 'get',
                url: options.killswitch?.enabled ? `${options.killswitch.instance}${consts_1.default.killswitch.sheri}?endpoint=${options.endpoint}${options.apikey ? `&apikey=${options.apikey}` : ""}` : `${consts_1.default.direct.sheri}/${options.endpoint}?format=json`,
                headers: {
                    "User-Agent": options.useragent,
                    // @ts-ignore
                    ...(options.apikey ? {
                        // @ts-ignore
                        "Authorization": options.apikey
                    } : {})
                }
            });
            return sherireq.data;
        case 'floofy':
            let floofyreq = await axios_1.default({
                method: 'get',
                url: options.killswitch?.enabled ? `${options.killswitch.instance}${consts_1.default.killswitch.floofy}` : `${consts_1.default.direct.floofy}/yiff`,
                headers: {
                    "User-Agent": options.useragent,
                }
            });
            return floofyreq.data;
        case 'shibe':
            let shibereq = await axios_1.default({
                method: 'get',
                url: options.killswitch?.enabled ? `${options.killswitch.instance}${consts_1.default.killswitch.shibe}?animal=${options.animal}&limit=${options.limit}` :
                    `${consts_1.default.direct.shibe}/${options.animal}?count=${options.limit}&urls=true&httpsUrls=true`,
                headers: {
                    "User-Agent": options.useragent,
                }
            });
            return shibereq.data;
        case 'fox':
            let randomfoxreq = await axios_1.default({
                method: 'get',
                url: options.killswitch?.enabled ? `${options.killswitch.instance}${consts_1.default.killswitch.fox}` : consts_1.default.direct.fox,
                headers: {
                    "User-Agent": options.useragent,
                }
            });
            return randomfoxreq.data;
        case 'thaldrin':
            let thaldrinreq = await axios_1.default({
                method: 'get',
                url: options.killswitch?.enabled ? `${options.killswitch.instance}${consts_1.default.killswitch.thaldrin}?endpoint=${options.endpoint}${options.apikey ? `&apikey=${options.apikey}` : ""}` : `${consts_1.default.direct.thaldrin}/categories/${options.endpoint}`,
                headers: {
                    "User-Agent": options.useragent,
                    // @ts-ignore
                    ...(options.apikey ? {
                        // @ts-ignore
                        "Authorization": options.apikey
                    } : {})
                }
            });
            return thaldrinreq.data;
        default:
            return {
                success: false,
                message: "No URL provided"
            };
    }
}
exports.default = request;
