// e621/build/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Custom = exports.Dev = exports.YiffyAPI = exports.WikiPageHistory = exports.WikiPage = exports.User = exports.Takedown = exports.TagHistory = exports.Tag = exports.PostSet = exports.PostHistory = exports.PostApproval = exports.Post = exports.PostFlag = exports.PoolHistory = exports.Pool = exports.NoteHistory = exports.Note = exports.Blip = exports.ArtistHistory = exports.Artist = exports.AuthenticatedUser = exports.APIError = void 0;
const RequestHandler_1 = __importDefault(require("./util/RequestHandler"));
// Modules
const Artists_1 = __importDefault(require("./modules/Artists"));
const Blips_1 = __importDefault(require("./modules/Blips"));
const Notes_1 = __importDefault(require("./modules/Notes"));
const Pools_1 = __importDefault(require("./modules/Pools"));
const Posts_1 = __importDefault(require("./modules/Posts"));
const PostFlags_1 = __importDefault(require("./modules/PostFlags"));
const PostSets_1 = __importDefault(require("./modules/PostSets"));
const Tags_1 = __importDefault(require("./modules/Tags"));
const Takedowns_1 = __importDefault(require("./modules/Takedowns"));
const UserFeedback_1 = __importDefault(require("./modules/UserFeedback"));
const Users_1 = __importDefault(require("./modules/Users"));
const WikiPages_1 = __importDefault(require("./modules/WikiPages"));
// Structures
const Artist_1 = __importDefault(require("./structures/Artist"));
exports.Artist = Artist_1.default;
const ArtistHistory_1 = __importDefault(require("./structures/ArtistHistory"));
exports.ArtistHistory = ArtistHistory_1.default;
const Blip_1 = __importDefault(require("./structures/Blip"));
exports.Blip = Blip_1.default;
const Note_1 = __importDefault(require("./structures/Note"));
exports.Note = Note_1.default;
const NoteHistory_1 = __importDefault(require("./structures/NoteHistory"));
exports.NoteHistory = NoteHistory_1.default;
const AuthenticatedUser_1 = __importDefault(require("./structures/AuthenticatedUser"));
exports.AuthenticatedUser = AuthenticatedUser_1.default;
const Pool_1 = __importDefault(require("./structures/Pool"));
exports.Pool = Pool_1.default;
const PoolHistory_1 = __importDefault(require("./structures/PoolHistory"));
exports.PoolHistory = PoolHistory_1.default;
const PostFlag_1 = __importDefault(require("./structures/PostFlag"));
exports.PostFlag = PostFlag_1.default;
const Post_1 = __importDefault(require("./structures/Post"));
exports.Post = Post_1.default;
const PostApproval_1 = __importDefault(require("./structures/PostApproval"));
exports.PostApproval = PostApproval_1.default;
const PostHistory_1 = __importDefault(require("./structures/PostHistory"));
exports.PostHistory = PostHistory_1.default;
const PostSet_1 = __importDefault(require("./structures/PostSet"));
exports.PostSet = PostSet_1.default;
const Tag_1 = __importDefault(require("./structures/Tag"));
exports.Tag = Tag_1.default;
const Takedown_1 = __importDefault(require("./structures/Takedown"));
exports.Takedown = Takedown_1.default;
const TagHistory_1 = __importDefault(require("./structures/TagHistory"));
exports.TagHistory = TagHistory_1.default;
const User_1 = __importDefault(require("./structures/User"));
exports.User = User_1.default;
const WikiPage_1 = __importDefault(require("./structures/WikiPage"));
exports.WikiPage = WikiPage_1.default;
const WikiPageHistory_1 = __importDefault(require("./structures/WikiPageHistory"));
exports.WikiPageHistory = WikiPageHistory_1.default;
const package_json_1 = __importDefault(require("../package.json"));
// note for future reference (browser compatibility?)
// POST with "_method=*" can be used in place of PUT/PATCH/DELETE
class E621 {
    static YiffyAPI;
    static Dev;
    static Custom;
    POST_LIMIT_PER_REQUEST = 320;
    request = new RequestHandler_1.default(this);
    artists = new Artists_1.default(this);
    blips = new Blips_1.default(this);
    notes = new Notes_1.default(this);
    pools = new Pools_1.default(this);
    posts = new Posts_1.default(this);
    postFlags = new PostFlags_1.default(this);
    postSets = new PostSets_1.default(this);
    tags = new Tags_1.default(this);
    takedowns = new Takedowns_1.default(this);
    userFeedback = new UserFeedback_1.default(this);
    users = new Users_1.default(this);
    wikiPages = new WikiPages_1.default(this);
    auth;
    options;
    constructor(options) {
        if (!options)
            options = {};
        Object.defineProperties(this, {
            auth: {
                get() {
                    return !this.options.authUser || !this.options.authKey ? null : `Basic ${Buffer.from(`${this.options.authUser}:${this.options.authKey}`).toString("base64")}`;
                },
                configurable: false,
                enumerable: false
            },
            options: {
                value: {
                    authUser: options.authUser ?? null,
                    authKey: options.authKey ?? null,
                    userAgent: options.userAgent ?? `E621/${package_json_1.default.version} (https://github.com/DonovanDMC/E621${!options.authUser ? "" : `; "${options.authUser}"`})`,
                    requestTimeout: options.requestTimeout ?? 30
                },
                configurable: false,
                enumerable: false,
                writable: false
            }
        });
        this.setInstance({});
    }
    setInstance(opt) {
        if (!opt.host)
            opt.host = "e621.net";
        opt.ssl = opt.ssl ?? ["e621.net", "yiff.rest"].includes(opt.host);
        if (!opt.port)
            opt.port = opt.ssl ? 443 : 80;
        if (!opt.imageReconstructionType) {
            switch (opt.host) {
                case "e621.net":
                    opt.imageReconstructionType = "e621";
                    break;
                case "yiff.rest":
                    opt.imageReconstructionType = "yiffy";
                    break;
                case "e621.local":
                case "e621ng.local":
                case "localhost":
                    opt.imageReconstructionType = "dev";
                    break;
                default: {
                    opt.imageReconstructionType = null;
                    if (!opt.reconstructStaticURL)
                        process.emitWarning("You specified a host that wasn't in our known list (e621.net, yiff.rest, e621.local) and didn't also specify imageReconstructionType or reconstructStaticURL. This WILL cause an error if we see any null file urls.");
                }
            }
        }
        Object.assign(this.options, {
            instanceHost: opt.host,
            instancePort: opt.port,
            instanceSSL: opt.ssl,
            reconstructStaticURL: opt.reconstructStaticURL || null,
            imageReconstructionType: opt.imageReconstructionType || "hierarchy"
        });
    }
}
exports.default = E621;
class YiffyAPI extends E621 {
    constructor(options) {
        super(options);
        this.setInstance({
            host: "yiff.rest",
            port: 443,
            ssl: true
        });
    }
}
exports.YiffyAPI = YiffyAPI;
class Dev extends E621 {
    constructor(options, ssl = false) {
        super(options);
        this.setInstance({
            host: "localhost",
            port: 3000,
            ssl
        });
    }
}
exports.Dev = Dev;
class Custom extends E621 {
    constructor(options, instanceOptions) {
        super(options);
        this.setInstance(instanceOptions);
    }
}
exports.Custom = Custom;
// due to extending the main class, neither of these (E621 class & sub-class) can happen "first", when both need to
E621.YiffyAPI = YiffyAPI;
E621.Dev = Dev;
E621.Custom = Custom;
module.exports = E621;
var RequestHandler_2 = require("./util/RequestHandler");
Object.defineProperty(exports, "APIError", { enumerable: true, get: function () { return RequestHandler_2.APIError; } });
