// e621/build/src/structures/Post.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Post {
    main;
    id;
    created_at;
    updated_at;
    file;
    preview;
    sample;
    score;
    tags;
    locked_tags;
    change_seq;
    flags;
    rating;
    fav_count;
    sources;
    pools;
    relationships;
    approver_id;
    uploader_id;
    description;
    comment_count;
    is_favorited;
    has_notes;
    duration;
    constructor(main, info) {
        Object.assign(this, info);
        Object.defineProperty(this, "main", {
            value: main,
            configurable: false,
            enumerable: false,
            writable: false
        });
        // blacklisted or deleted
        if (this.file.url === null)
            this.file.url = this.flags.deleted ? this.main.request.deletedImageURL : this.main.request.constructURL(this.file.md5, "original", this.file.ext);
        if (this.preview.url === null)
            this.preview.url = this.flags.deleted ? this.main.request.deletedImageURL : this.main.request.constructURL(this.file.md5, "preview", this.file.ext);
        if (this.sample.has && this.sample.url === null)
            this.sample.url = this.flags.deleted ? this.main.request.deletedImageURL : this.main.request.constructURL(this.file.md5, "sample", this.file.ext);
    }
    /**
     * Get the user object for creator of this post
     *
     * @returns {Promise<User | null>}
     */
    async getUploader() { return this.main.users.get.call(this.main.users, this.uploader_id); }
    /**
     * Get the user object for approver of this post (will be null if self approved)
     *
     * @returns {Promise<User | null>}
     */
    async getApprover() { return this.approver_id === null ? null : this.main.users.get.call(this.main.users, this.approver_id); }
    /**
     * Add this post to a set
     *
     * * Requires Authentication
     *
     * @param {number} id - the id of the set to add this post to
     * @returns {Promise<PostSet>}
     */
    async addToSet(id) {
        this.main.request.authCheck("Post#addToSet");
        return this.main.postSets.addPost.call(this.main.postSets, id, this.id);
    }
    /**
     * Remove this post from a set
     *
     * * Requires Authentication
     *
     * @param {number} id - the id of the set to remove this post from
     * @returns {Promise<PostSet>}
     */
    async removeFromSet(id) {
        this.main.request.authCheck("Post#removeFromSet");
        return this.main.postSets.removePost.call(this.main.postSets, id, this.id);
    }
    /**
     * Add this post to a pool
     *
     * * Requires Authentication
     *
     * @param {number} id - the id of the pool to add this post to
     * @returns {Promise<Pool>}
     */
    async addToPool(id) {
        this.main.request.authCheck("Post#addToPool");
        return this.main.pools.addPost.call(this.main.postSets, id, this.id);
    }
    /**
     * Remove this post from a pool
     *
     * * Requires Authentication
     *
     * @param {number} id - the id of the pool to remove this post from
     * @returns {Promise<Pool>}
     */
    async removeFromPool(id) {
        this.main.request.authCheck("Post#removeFromPool");
        return this.main.pools.removePost.call(this.main.pools, id, this.id);
    }
    /**
     * Modify a post
     *
     * * Requires Authentication
     *
     * @param {object} options
     * @param {string} [options.edit_reason] - the reason for the edit
     * @param {(Array<string> | string)} [options.add_tags] - the tags to add to the post
     * @param {(Array<string> | string)} [options.remove_tags] - the tags to remove from the post
     * @param {(Array<string> | string)} [options.add_sources] - the sources to add to the post
     * @param {(Array<string> | string)} [options.remove_sources] - the sources to remove from the post
     * @param {Ratings} [options.rating] - the rating for the post
     * @param {string} [options.description] - the description of the post
     * @param {number} [options.parent_id] - the parent of the post
     * @param {boolean} [options.has_embedded_notes] - I don't know what this is, api docs list it though
     * @param {boolean} [options.rating_locked] - upload as rating locked (requires privileged)
     * @param {boolean} [options.note_locked] - upload as rating locked (requires janitor)
     * @param {boolean} [options.status_locked] - upload as rating locked (requires admin)
     * @param {boolean} [options.hide_from_anonymous] - upload as rating locked (requires admin)
     * @param {boolean} [options.hide_from_search] - upload as rating locked (requires admin)
     * @param {string} [options.background_color] - the background color of the post (requires janitor)
     * @param {(Array<string> | string)} [options.lockedTags] - tags to lock on the post (requires admin)
     * @returns {Promise<Post>}
     */
    async modify(options) {
        this.main.request.authCheck("Post#modify");
        if (!options)
            throw new Error("options is required in Post#modify");
        return this.main.posts.modify.call(this.main.posts, this.id, options);
    }
    /**
     * Vote on this post
     *
     * * Requires Authentication
     *
     * @param {boolean} up - if the vote should be up or down
     * @returns {Promise<PostVoteResult>}
     */
    async vote(up) {
        this.main.request.authCheck("Post#vote");
        return this.main.posts.vote.call(this.main.posts, this.id, up);
    }
    /**
     * Favorite this post
     *
     * * Requires Authentication
     *
     * @returns {Promise<Post>}
     */
    async favorite() {
        this.main.request.authCheck("Post#favorite");
        return this.main.users.addFavorite.call(this.main.users, this.id);
    }
    /**
     * Unfavorite this post
     *
     * * Requires Authentication
     *
     * @returns {Promise<Post>}
     */
    async unfavorite() {
        this.main.request.authCheck("Post#unfavorite");
        return this.main.users.removeFavorite.call(this.main.users, this.id);
    }
}
exports.default = Post;
