// e621/build/src/structures/Tag.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Tag {
    main;
    id;
    name;
    created_at;
    updated_at;
    creator_id;
    post_count;
    category;
    related_tags;
    related_tags_updated_at;
    is_locked;
    constructor(main, info) {
        Object.assign(this, info);
        Object.defineProperty(this, "main", {
            value: main,
            configurable: false,
            enumerable: false,
            writable: false
        });
    }
    /**
     * Get posts with this tag (and others, if specified)
     *
     * @param {object} [options]
     * @param {(Array<string> | string)} [options.tags] - narrow the search by specific tags
     * @param {(number |`${"" | "a" | "b"}${number}`)} [options.page] - number for exact page, a${number} posts after ${number}, b${number} posts before ${number}
     * @param {number} [options.limit] - limit the maximum amount of results returned
     *
     * @returns {Promise<Array<Post>>}
     */
    async getPosts(options) { return this.main.posts.search.call(this.main.posts, { ...(options ?? {}), tags: [this.name, ...(options?.tags ?? [])] }); }
    /**
     * Fetch all posts with this tag (and extras, if specified)
     *
     * @param {Array<string>} [extraTags] - the additional tags for searching
     * @param {number} [lastID] - internal use only
     * @param {Array<Post>} [lastPosts] - internal use only
     * @returns
     */
    async getAllPosts(extraTags, lastID, lastPosts = []) {
        const posts = await this.main.posts.search({
            tags: [this.name, ...(extraTags ?? [])],
            page: lastID === undefined ? undefined : `b${lastID}`,
            limit: this.main.POST_LIMIT_PER_REQUEST
        });
        lastPosts.push(...posts);
        if (posts.length === this.main.POST_LIMIT_PER_REQUEST)
            return this.getAllPosts(extraTags, lastID, lastPosts);
        else
            return lastPosts;
    }
    /**
     * modify this tag
     *
     * * Requires Authentication
     *
     * @param {object} options
     * @param {number} [options.category] - the category of the tag
     * @param {string} [options.locked] - if the tag is locked (requires moderator)
     * @returns {Promise<Tag>}
     */
    async modify(options) {
        this.main.request.authCheck("Pool#modify");
        if (!options)
            throw new Error("options is required in Tags#modify");
        return this.main.tags.modify.call(this.main.tags, this.id, options);
    }
    /**
     * Search this tag's history
     *
     * @param {object} [options]
     * @param {number} [options.user] - page of results to get
     * @param {number} [options.user_id] - page of results to get
     * @param {number} [options.page] - page of results to get
     * @param {number} [options.limit] - limit the maximum amount of results returned
     * @param
     * @returns {Promise<Array<TagHistory>>}
     */
    async getHistory(options) {
        return this.main.tags.searchHistory.call(this.main.tags, { id: this.id, ...options });
    }
}
exports.default = Tag;
